comps/cores/common/component.py

```diff
@@ -0,0 +1,155 @@
+
+        Returns:
+            dict: A dictionary containing the component's metadata.
+        """
+        return {
+            "name": self.name,
+            "type": self.type,
+            "description": self.description,
+            "config": self.config,
+        }
+
+    def update_config(self, key: str, value):
+        """Updates a configuration parameter for the component.
+
+        Args:
+            key (str): The configuration parameter's key.
+            value: The new value for the configuration parameter.
+        """
+        self.config[key] = value
+
+    @abstractmethod
+    def check_health(self) -> bool:
+        """Checks the health of the component.
+
+        Returns:
+            bool: True if the component is healthy, False otherwise.
+        """
+        pass
+
+    @abstractmethod
+    def invoke(self, *args, **kwargs):
+        """Invoke service accessing using the component.
+
+        Args:
+            *args: Positional arguments.
+            **kwargs: Keyword arguments.
+
+        Returns:
+            Any: The result of the service accessing.
+        """
+        pass
+
+    def __repr__(self):
+        """Provides a string representation of the component for debugging and logging purposes.
+
+        Returns:
+            str: A string representation of the OpeaComponent instance.
+        """
+        return f"OpeaComponent(name={self.name}, type={self.type}, description={self.description})"
+
+
+class OpeaComponentController(ABC):
+    """The OpeaComponentController class serves as the base class for managing and orchestrating multiple
+    instances of components of the same type. It provides a unified interface for routing tasks,
+    registering components, and dynamically discovering available components.
+
+    Attributes:
+        components (dict): A dictionary to store registered components by their unique identifiers.
+    """
+
+    def __init__(self):
+        """Initializes the OpeaComponentController instance with an empty component registry."""
+        self.components = {}
+        self.active_component = None
+
+    def register(self, component):
+        """Registers an OpeaComponent instance to the controller.
+
+        Args:
+            component (OpeaComponent): An instance of a subclass of OpeaComponent to be managed.
+
+        Raises:
+            ValueError: If the component is already registered.
+        """
+        if component.name in self.components:
+            raise ValueError(f"Component '{component.name}' is already registered.")
+        self.components[component.name] = component
+
+    def discover_and_activate(self):
+        """Discovers healthy components and activates one.
+
+        If multiple components are healthy, it prioritizes the first registered component.
+        """
+        for component in self.components.values():
+            if component.check_health():
+                self.active_component = component
```

Comment:
If there are multiple healthy components, the first one will be asigned to self.activate_component?

Reply:
yes, currently we assume only one component is available.